### DISCUSSION TITLE
Make Entity class Serializable?

### DISCUSSION THREAD

**Discussion by chengenzhao** [AlmasB/FXGL#1220] [Ideas]
Hi:

We are working on save & load service/handlers
and 1 requirement is some people hope to save the instant state of the game world
in order to do that, we hope to save the entities in the game world
but current Entity class doesn't implements Serializable
thus it can not be saved directly
so the questions is: Is it a good idea to make entity class Serializable?
Seems like no harm but I am not sure about the performance etc.

Thanks

**Comment by AlmasB** [comment]
Entity is a list of components. A ViewComponent is a required component, backed by a Node, which is not Serializable. Therefore, Entity cannot be Serializable.

The currently supported approach is to serialise components via the `SerializableComponent` interface. Here's an example of TransformComponent:

```
    override fun write(bundle: Bundle) {
        bundle.put("propX", x)
        bundle.put("propY", y)
        bundle.put("angle", angle)
        bundle.put("scaleX", scaleX)
        bundle.put("scaleY", scaleY)
        bundle.put("scaleOriginX", scaleOrigin.x)
        bundle.put("scaleOriginY", scaleOrigin.y)
        bundle.put("rotationOriginX", rotationOrigin.x)
        bundle.put("rotationOriginY", rotationOrigin.y)
    }

    override fun read(bundle: Bundle) {
        setPosition(bundle.get("propX"), bundle.get("propY"))
        scaleOrigin = Point2D(bundle.get("scaleOriginX"), bundle.get("scaleOriginY"))
        rotationOrigin = Point2D(bundle.get("rotationOriginX"), bundle.get("rotationOriginY"))
        angle = bundle.get("angle")
        scaleX = bundle.get("scaleX")
        scaleY = bundle.get("scaleY")
    }
```

I never finished the design of Entity serialisation this way, as there wasn't a ready-to-go demo to support it, but ultimately, it should look like this (pseudo code):

```
Entity e;
List<Bundle> components = e.getComponents() -> .write()

Bundle b = new Bundle("playerEntity");
b.putAll(components);

// here "b" is Serializable, so can be used with JVM serialization
```
